src/libraries/System.Text.Json/docs/ReferenceHandling_spec.md

```diff
@@ -0,0 +1,810 @@
+
+Output:
+```json
+{
+    "$id": "1",
+    "$values": [
+        {
+            "$id": "2",
+            "Name": "Angela"
+        },
+        {
+            "$id": "3",
+            "Name": "Angela"
+        }
+    ]
+}
+```
+
+* **Deserialization**:
+The deserializer will throw when it reads `$ref` within a property that matches to a value type (such as a struct) and `ReferenceHandling.Preserve` is set. 
+
+Example:
+```cs
+public static void DeserializeStructs()
+{
+    string json = @"
+    {
+        ""$id"": ""1"",
+        ""$values"": [
+            {
+                ""$id"": ""2"",
+                ""Name"": ""Angela""
+            },
+            {
+                ""$ref"": ""2""
+            }
+        ]
+    }";
+
+    var options = new JsonSerializerOptions 
+    { 
+        ReferenceHandling = ReferenceHandling.Preserve
+    };
+
+    List<EmployeeStruct> root = JsonSerializer.Deserialize<List<EmployeeStruct>>(json, options);
+    // Throws JsonException.
+}
+```
+
+In other words, having a `$ref` property in a struct, is never emitted by the serializer and read such thing (by manually changing the JSON payload) is not supported by the deserializer.
+
+## Interaction with JsonPropertyNameAttribute
+Having the following class:
+
+```cs
+private class EmployeeAnnotated
+{
+    [JsonPropertyName("$id")]
+    public string Identifier { get; set; }
+    [JsonPropertyName("$ref")]
+    public string Reference { get; set; }
+    [JsonPropertyName("$values")]
+    public List<EmployeeAnnotated> Values { get; set; }
+
+    public string Name { get; set; }
+}
+```
+
+Either on Serialization or Deserialization:
+
+```cs
+public static void DeSerializeWithPreserve()
+{
+    var root = new EmployeeAnnotated();
+    var opts = new JsonSerializerOptions
+    {
+        ReferenceHandling = ReferenceHandling.Preserve
+    };
+
+    //Throw JsonException - PropertyName cannot start with '$' when Preserve References is enabled.
+    string json = JsonSerializer.Serialize(root, opts);
+    
+    //Also throws the same exception.
+    EmployeeAnnotated obj = JsonSerializer.Deserialize(json, opts);
+}
+```
+
+# Future
+Things that may build on top based on customer feedback:
+
+* (De)Serialize can define its own Preserve References Handling behavior each (i.e: you could opt-out form Preserve Reference on serialization but opt-in for read them on deserialization).
+
+* Expose a `ReferenceResolver` to override the logic that preserves references (Create your own implementation of a reference resolver).
+
+* Expose the `ReferenceResolver` in Converters to have access to the map of references.
+
+* Create `JsonReferenceHandlingAttribute` to enable to annotate properties and classes with their own isolated ReferenceHandling behavior (I cut-off the support for this due the constant checking for attributes was causing too much perf overhead on the main path, maybe we can try moving the attribute check to the warm-up method to reduce the runtime increase).
```

Comment:
```suggestion
* Create `JsonReferenceHandlingAttribute` to enable annotating properties and classes with their own isolated `ReferenceHandling` behavior (I am deferring this feature because the constant checking for attributes was causing too much perf overhead on the main path, but maybe we can try moving the attribute check to the warm-up method to reduce the runtime increase).
```